lib/using_restricted_array.rb

```diff
@@ -6,58 +6,126 @@
 
 # Calculates the length of the restricted array. All values are integers.
 # The restricted_array is terminated by 'nil' i.e. array[length] = nil
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def length(array)
-  raise NotImplementedError
+  count = 0
+  while array[count] != nil
+    count += 1
+  end
+  return count
 end
 
 # Prints each integer values in the array
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def print_array(array)
-  raise NotImplementedError
+  count = 0
+  while array[count] != nil
+    print "#{array[count]} "
+    count += 1
+  end
 end
 
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def search(array, length, value_to_find)
-  raise NotImplementedError
+  count = 0
+  length.times do
+    if array[count] == value_to_find
+      return true
+    end
+    count += 1
+  end
+  return false
 end
 
 # Finds and returns the largest integer value the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def find_largest(array, length)
-  raise NotImplementedError
+  count = 0
+  largest_value = 0
+  length.times do
+    if array[count] > largest_value
+      largest_value = array[count]
+    end
+    count += 1
+  end
+  return largest_value
 end
 
 # Finds and returns the smallest integer value in the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def find_smallest(array, length)
-  raise NotImplementedError
+  count = 0
+  smallest_value = array[0]
+  length.times do
+    if array[count] < smallest_value
+      smallest_value = array[count]
+    end
+    count += 1
+  end
+  return smallest_value
 end
 
 # Reverses the values in the integer array in place
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n2)
```

Comment:
Your two loops are not nested, and instead sequential, so the time complexity is O(n)

You do build another array so the space complexity is also O(n)